virtualmachine/vsphere/util.go

```diff
@@ -455,15 +458,61 @@ func reconfigureNetworks(vm *VM) ([]types.BaseVirtualDeviceConfigSpec, error) {
 		return nil, err
 	}
 
-	// for all the mappings create a nic and add network backing info to it
-	// create deviceSpec array from  the mapping
-	for _, mapping := range networkMapping {
-		spec, err := addNetworkDeviceSpec(mapping.Network, mapping.Name)
-		if err != nil {
-			return nil, err
+	devices, err := vmObj.Device(vm.ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	// Modify existing networks in template with provided networks list
+	for _, device := range devices {
+		switch device.(type) {
+		case *types.VirtualE1000, *types.VirtualE1000e, *types.VirtualVmxnet3:
+			if len(vm.Networks) == 0 {
+				// Remove extra networks
+				spec := &types.VirtualDeviceConfigSpec{
+					Operation: types.VirtualDeviceConfigSpecOperationRemove,
+					Device:    device,
+				}
+				deviceSpecs = append(deviceSpecs, spec)
+				continue
+			}
+
+			// Edit device
+			nw, vm.Networks = vm.Networks[0], vm.Networks[1:]
```

Comment:
This feels a little unreadable. The following feels easier to read:

    nws_len := len(vm.Networks)
    devs_len := len(devices)
    
    if devs_len < nws_len
      // generate all the remove specs here
      // on vm.Networks[devs_len:]
    end
    
    // generate all the edit specs on devices[0:devs_len] (or devices[0:devs_len-1] - check syntax)

There's also less garbage to collect.